ring-jetty-adapter/src/ring/adapter/jetty.clj

```diff
@@ -28,18 +28,48 @@
           (servlet/update-servlet-response response response-map)
           (.setHandled base-request true))))))
 
-(defn- ^AbstractHandler async-proxy-handler [handler timeout]
+(defn- async-jetty-raise
+  [^AsyncContext context ^HttpServletResponse response]
+  (fn [^Throwable exception]
+    (.sendError ^HttpServletResponse response
+                500
+                (.getMessage exception))
+    (.complete ^AsyncContext context)))
+
+(defn- async-jetty-respond
+  [^AsyncContext context ^HttpServletResponse response]
+  (fn [response-map]
+    (servlet/update-servlet-response response context response-map)))
+
+(defn- async-timeout-listener
+  [request context response handler]
+  (proxy [AsyncListener] []
+    (onTimeout [^AsyncEvent _]
+      (handler
+       (servlet/build-request-map request)
+       (async-jetty-respond context response)
+       (async-jetty-raise context response)))
+    (onComplete [^AsyncEvent _])
+    (onError [^AsyncEvent _])
+    (onStartAsync [^AsyncEvent _])))
+
+(defn- ^AbstractHandler async-proxy-handler
+  [handler timeout timeout-handler]
   (proxy [AbstractHandler] []
     (handle [_ ^Request base-request ^HttpServletRequest request ^HttpServletResponse response]
       (let [^AsyncContext context (.startAsync request)]
         (.setTimeout context timeout)
+        (when timeout-handler
+          (.addListener
+           context
+           (async-timeout-listener request
```

Comment:
This could be more neatly formatted as:

```clojure
          (.addListener
           context
           (async-timeout-listener request context response timeout-handler)))
```